new/resource-actions.md

```diff
@@ -0,0 +1,199 @@
+
+## Specification
+
+To create an action on a resource, users specify `action <action> do ... end`.
+The recipe in between the do ... end block will be run when the action is
+performed.
+
+```ruby
+class MyResource < Chef::Resource::LWRPBase
+  attribute :path, name_attribute: true
+  attribute :content
+  attribute :mode
+
+  action :create do
+    ruby_block "Update content of #{path}" do
+      only_if { content != IO.read(path) }
+      block { IO.write(path, content) }
+    end
+
+    execute "Update mode of #{path}" do
+      not_if { File.stat(path).mode == mode }
+      command "chmod #{mode} #{path}"
+    end
+  end
+end
+```
+
+There is no change to providers or the way we run actions.  This is just a way
+of defining a provider without having to type or think about the word.  Providers
+do not go away, so much as become a secondary concept, a sort of house for the
+primary concept of actions.
+
+`action` will work for any resource extending from `Chef::Resource`.
+
+### Action details
+
+#### Action recipe execution
+
+The action recipe works like a Chef run, in that it is an isolated resource
+collection and runs actions and notifications the same way as a top level Chef
+run.
+
+- The entire action recipe is compiled (just like a normal recipe) before
+  converging.
+- This compile happens when the action is *run*, not when the parent resource
+  is declared.
+- Delayed and immediate notifications are local to the action.
+
+This is equivalent to `use_inline_providers` in a Provider, and in fact is
+implemented that way.  We just call out the behavior here to make it clear what will happen.
+
+#### Action composition and inheritance
+
+The action recipes are all added to a `Provider` class which inherits from the
+parent resource's provider class.  You may call `super` to call the parent
+class's version of the current action.
+
+```ruby
+class MyFile < Chef::Resource::File
+  action :create do
+    puts "I am a file yo"
+    super
+  end
+end
+```
+
+### Inline resource output
+
+When `use_inline_resources` runs, the output can be much more verbose than the
+user expects.  Particularly, they see a number of resources and actions that they didn't declare:
+
+```
+* x[blah] action blah
+  * file[/Users/jkeiser/x.txt] action create (up to date)
+   (up to date)
+```
+
+We suggest not actually showing sub-resources as separate things in output:
+
+```
+* x[blah] action blah (up to date)
+```
+
+An "update" would look like this (again, removes the extra nesting):
+
+```
+* x[blah] action blah
+  - create new file /Users/jkeiser/x.txt
+  - update content in file /Users/jkeiser/x.txt from none to a591a6
+  --- /Users/jkeiser/x.txt	2015-04-29 08:15:14.000000000 -0700
+  +++ /Users/jkeiser/.x.txt20150429-56731-6xk0lx	2015-04-29 08:15:14.000000000 -0700
+  @@ -1 +1,2 @@
+  +Hello World
+```
+
+The principle is: the user doesn't see resources they don't type.
```

Comment:
I don't like this particular bikeshed at all.  I read what the LWRPs are doing all the time and it helps me make sense of what is going on inside of them, and I don't find it confusing at all.